app/assets/javascripts/discourse/app/lib/utilities.js

```diff
@@ -421,34 +422,40 @@ export function postRNWebviewMessage(prop, value) {
   }
 }
 
-const CODE_BLOCKS_REGEX =
-  /^(  |\t).*|`[^`]+`|^```[^]*?^```|\[code\][^]*?\[\/code\]/gm;
-//|    ^     |   ^   |      ^      |           ^           |
-//     |         |          |                  |
-//     |         |          |       code blocks between [code]
-//     |         |          |
-//     |         |          +--- code blocks between three backticks
-//     |         |
-//     |         +----- inline code between backticks
-//     |
-//     +------- paragraphs starting with 2 spaces or tab
+function pickMarker(text) {
+  // Uses the private use area (U+E000 to U+F8FF) to find a character that
+  // is not present in the text. This character will be used as a marker in
+  // place of the caret.
+  for (let code = 0xe000; code <= 0xf8ff; ++code) {
+    const char = String.fromCharCode(code);
+    if (!text.includes(char)) {
+      return char;
+    }
+  }
+  return null;
+}
 
-const OPEN_CODE_BLOCKS_REGEX = /^(  |\t).*|`[^`]+|^```[^]*?|\[code\][^]*?/gm;
+function findToken(tokens, marker, level = 0) {
+  if (level > 50) {
+    return null;
+  }
+  const token = tokens.find((t) => (t.content ?? "").includes(marker));
+  return token?.children ? findToken(token.children, marker, level + 1) : token;
+}
 
-export function inCodeBlock(text, pos) {
-  let end = 0;
-  for (const match of text.matchAll(CODE_BLOCKS_REGEX)) {
-    end = match.index + match[0].length;
-    if (match.index <= pos && pos <= end) {
-      return true;
-    }
+const CODE_TOKEN_TYPES = ["code_inline", "code_block", "fence"];
+
+export async function inCodeBlock(text, pos) {
+  const marker = pickMarker(text);
+  if (!marker) {
+    return false;
```

Comment:
If we can't find a unique character to use as a marker, then we bail out. We probably have bigger problems than showing the autocomplete 😅